lib/pbio/include/pbio/os.h

```diff
@@ -0,0 +1,282 @@
+
+/**
+ * A process.
+ */
+struct _pbio_os_process_t {
+    /**
+     * Pointer to the next process in the list.
+     */
+    pbio_os_process_t *next;
+    /**
+     * State of the protothread.
+     */
+    pbio_os_state_t state;
+    /**
+     * The protothread.
+     */
+    pbio_os_process_func_t func;
+    /**
+     * Context passed on each call to the protothread.
+     */
+    void *context;
+    /**
+     * Most recent result of running one iteration of the protothread.
+     */
+    pbio_error_t err;
+};
+
+/**
+ * Initialize a protothread.
+ *
+ * Initializes a protothread. Initialization must be done prior to
+ * starting to execute the protothread.
+ *
+ * @param [in]  state    Protothread state.
+ */
+#define ASYNC_INIT(state)   PB_LC_INIT(state)
+
+/**
+ * Declare the start of a protothread inside the C function
+ * implementing the protothread.
+ *
+ * This macro is used to declare the starting point of a
+ * protothread. It should be placed at the start of the function in
+ * which the protothread runs. All C statements above the ASYNC_BEGIN()
+ * invocation will be executed each time the protothread is scheduled.
+ *
+ * @param [in]  state    Protothread state.
+ */
+#define ASYNC_BEGIN(state) { char do_yield_now = 0; if (do_yield_now) {;} PB_LC_RESUME(state)
+
+/**
+ * Declare the end of a protothread and returns with given code.
+ *
+ * This macro is used for declaring that a protothread ends. It must
+ * always be used together with a matching ASYNC_BEGIN() macro.
+ *
+ * NB: In contrast to Contiki, this does not call ASYNC_INIT() before exiting.
+ *
+ * @param [in]  err    Error code to return.
+ */
+#define ASYNC_END(err) PB_LC_END(); return err; }
+
+/**
+ * Yields the protothread while the specified condition is true.
+ *
+ * @param [in]  state     Protothread state.
+ * @param [in]  condition The condition.
+ */
+#define AWAIT_WHILE(state, condition)            \
+    do {                                         \
+        PB_LC_SET(state);                        \
+        if (condition) {                         \
+            return PBIO_ERROR_AGAIN;             \
+        }                                        \
+    } while (0)
+
+/**
+ * Yields the protothread until the specified condition is true.
+ *
+ * @param [in]  state     Protothread state.
+ * @param [in]  condition The condition.
+ */
+#define AWAIT_UNTIL(state, condition)            \
+    do {                                         \
+        PB_LC_SET(state);                        \
+        if (!(condition)) {                      \
+            return PBIO_ERROR_AGAIN;             \
+        }                                        \
+    } while (0)
+
+/**
+ * Awaits a protothread until it is done.
+ *
+ * @param [in]  state      Protothread state.
+ * @param [in]  child      Protothread state of the child.
+ * @param [in]  statement  The statement to await.
+ */
+#define AWAIT(state, child, statement)          \
```

Comment:
Technically, this need to be an _expression_, not a _statement_. And the doc comment needs to say that the expression needs to evaluate to a `pbio_error_t`. Same applies to the `AWAIT_RACE` as well.